src/FwdState.cc

```diff
@@ -802,14 +808,25 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         Must(!Comm::IsConnOpen(answer.conn));
         answer.error.clear(); // preserve error for errorSendComplete()
     } else if (!Comm::IsConnOpen(answer.conn) || fd_table[answer.conn->fd].closing()) {
+        // We do not know exactly why the connection got closed, so we play it
+        // safe, allowing retries only for persistent (reused) connections
+        if (answer.reused) {
+            destinationReceipt = answer.conn;
+            assert(destinationReceipt);
+        }
         syncHierNote(answer.conn, request->url.host());
         closePendingConnection(answer.conn, "conn was closed while waiting for noteConnection");
         error = new ErrorState(ERR_CANNOT_FORWARD, Http::scServiceUnavailable, request, al);
+    } else {
+        assert(!error);
+        destinationReceipt = answer.conn;
+        assert(destinationReceipt);
+        // serverConn remains nil until syncWithServerConn()
     }
 
     if (error) {
         fail(error);
-        retryOrBail(); // will notice flags.dont_retry and bail
+        retryOrBail();
```

Comment:
FTR: The documentation for this comment removal is in unofficial commit 22f55cda04f2c7b15e8b2da83b1d3f52bba43c60 message, quoted here:

> Removed stale (and dangerously misleading) comment

> This branch changes did not make that comment stale -- FwdState may actually retry if "conn was closed while waiting for noteConnection", but the comment is in the way of correctly interpreting the related code that is being modified in this branch.